rs/replicated_state/src/replicated_state.rs

```diff
@@ -1044,6 +1048,23 @@ impl ReplicatedState {
     }
 }
 
+/// Converts a `CanisterInput` popped from a subnet input queue into a
+/// `CanisterMessage`.
+///
+/// As opposed to actual canister queues, subnet input queues should never hold
+/// responses.
+fn subnet_input_into_canister_message(input: CanisterInput) -> CanisterMessage {
+    match input {
+        CanisterInput::Ingress(ingress) => CanisterMessage::Ingress(ingress),
+        CanisterInput::Request(request) => CanisterMessage::Request(request),
+        CanisterInput::Response(_)
+        | CanisterInput::DeadlineExpired(_)
+        | CanisterInput::ResponseDropped(_) => {
+            unreachable!("Subnet input queues should never hold responses")
```

Comment:
Do we really need this `unreachable` here? It feels to me that this might not be the right place to enforce this. If it ever changes and execution starts sending requests from the management canister, this unreachable here will almost certainly be overlooked.

Reply:
I guess one could argue about `CanisterInput::Response(_)`: we could simply convert it into a `CanisterMessage` and return it.

But there isn't much we could do about a `DeadlineExpired` or `ResponseDropped`: getting to the right kind of callback and generating a matching reject response would be quite involved; and it would be dead code that is as likely to break if it ever goes live as if we simply panic instead. Plus, a panic would immediately catch any outgoing management canister request / incoming response. Whereas a slightly bungled generated reject response might easily make it onto mainnet.

Reply:
I see makes sense. Not sure I understand yet why we couldn't simply do the same as for canisters? Is it because we don't have dedicated call context manager for the subnet queue, where the concept of looking up callback ids for a response doesn't make sense?

If we go for this unreachable here, we should at least make sure we drop responses before attempting to induct [here](https://github.com/dfinity/ic/blob/master/rs/replicated_state/src/replicated_state.rs#L727). (The comment in the linked code suggests to assert but this would be dangerous as we are processing input from potentially dishonest subnets there, so we should just drop the response) as an additional line of defense. 

We should also make execution aware that we are codifying this implicit contract with this PR to see whether there are any objections @dsarlis.

Reply:
> Is it because we don't have dedicated call context manager for the subnet queue and thus the concept of looking up callback ids for timed out/shed responses doesn't make sense?

There exists in fact a `SubnetCallContextManager`, but it's a collection of `BTreeMaps` holding various types of DKG, HTTP, Bitcoin and other call contexts (and no callbacks). So in the future one may need to e.g. look inside one or more of the 8 kinds of call contexts to find callbacks, but it's just going to be a huge `match` statement.

> If we go for this unreachable here, we should at least make sure we drop responses before attempting to induct [here](https://github.com/dfinity/ic/blob/master/rs/replicated_state/src/replicated_state.rs?rgh-link-date=2024-09-12T11%3A29%3A44Z#L727).

Good point, will do.

> We should also make execution aware that we are codifying this implicit contract that the management canister will never send a request with this PR to see whether there are any objections.

I'll also ping them on Slack.

Reply:
Sent out #1471 to address this (as it's related to shed inbound responses; and this PR is already huge enough as it is). It simply produces an error when there's an attempt to enqueue a `Response` into a subnet input queue; which, in turn, results in a critical error in `StreamHandlerImpl`, nothing more.

Reply:
One alternative thought: maybe we don't even need to check whether the callback is present in the call context manager? Then we could treat both canister input and subnet input the same here. 

The idea came up because I thought about unwrapping the callback in `to_reject_response_above` and my thought was that I don't particularly like it because we are asserting for something that is cross component.

Reply:
Me must. The function takes a `CanisterInput` argument (potentially 	`DeadlineExpired` or `ResponseDropped`) and must convert it into a `CanisterMessage` (`Response`, in this case). Technically, nothing is currently checking the `respondent` or `deadline` against the `Callback` right now, so we could just put in some arbitrary values.

But I'd rather not introduce even more obscure sources of errors into the process.

> we are asserting for something that is cross component which is always dangerous and error prone.

It is not cross-component. It's all inside the `SystemState`. (Leaving aside the fact that `SystemState` will hand out a mutable reference to the `CallContextManager` within to anyone who bothers to ask for it.)

Reply:
> [...] But I'd rather not introduce even more obscure sources of errors into the process.

Oh, I missed that we are actually reading quite some stuff from the callback. I somehow assumed we'd just unwrap it here for the purpose of making sure that we actually have the callback around. Thanks for clarifying. I agree that we shouldn't just pass arbitrary values. 

> It is not cross-component. [...]

True. What I meant, though, is that the `CallContextManager` is owned by Execution whereas the queue related stuff is owned by `MessageRouting`.

Reply:
FWIW, I pulled another all-weekender and refactored `SystemState` to no longer hand out mutable references to (or into) the `CallContextManager`: #1498.

It doesn't yet fully protect about inconsistent mutations to `CallContextManager` vs `CanisterQueues` (nor can it do so trivially, as there can be a multi-round gap between e.g. popping an inbound response and unregistering its callback). But I'm quite sure that this should allow us to introduce invariants such as "don't pop another input message while the callback for the previously popped response is still there"; and to consolidate pushing an outbound requests and creating the corresponding callback into one single call. I guess that with enough work one could turn the `SystemState` into a state machine.